design/machine-remediation-request.md

```diff
@@ -0,0 +1,142 @@
+    * [Test Plan](#test-plan)
+    * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    * [Version Skew Strategy](#version-skew-strategy)
+  * [Alternatives](#alternatives)
+  * [References](#references)
+
+<!-- Added by: dhellmann, at: Wed Jun 26 16:08:57 EDT 2019 -->
+
+<!--te-->
+
+## Summary
+
+This document explains the implementation of the Machine Remediation Request controller and
+how it will work together with Machine Health Check controller to restore unhealthy machines.
+
+## Motivation
+
+To provide a mechanism to apply remediation strategy on an unhealthy node, that will be different by type and by a provider.
+
+### Goals
+
+1. To provide the interface that will make possible to apply different remediation strategy for each cloud or bare metal providers.
+2. To provide details regarding the progress of the remediation operation.
+
+### Non-Goals
+
+1. To provide a remediation strategy for each cloud or bare metal providers.
+2. To provide the integration with the **MachineHealthCheck** controller.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+It's common that high available environment should have a mechanism to restore unhealthy nodes and when delete and create a new node looks feasible solution for a cloud provider, in case of bare metal hosts with additional infrastructure configuration it looks like too destructive approach.
+To make possible to specify the remediation strategy for different failure cases and for nodes that run on different providers, this proposal introduces **MachineRemediationRequest** controller that should monitor all **MachineRemediationRequest** CR's and apply the remediation strategy according to details specified under the CR.
+When the **MachineRemediationRequest** will apply the remediation strategy, one who is initiating it will be the **MachineHealthCheck** controller. It will monitor nodes, and for each unhealthy node will create **MachineRemediationRequest** CR, if after the specified timeout the **MachineRemediationRequest** will not have the `remediationState` equal to `Succeeded`, the **MachineHealthCheck** controller will notify about the remediation failure, and can try to use different type of the remediation strategy.
+
+#### MachineRemediationRequest spec API
+
+This proposal introduces a new API type: **MachineRemediationRequest**, that makes possible to create simple **MachineRemediationRequest** object with the type of the remediation strategy for the specific machine.
+
+##### Example
+
+```yaml
+apiVersion: machineremediationrequest.metal3.io/v1alpha1
+kind: MachineRemediationRequest
+metadata:
+  name: mrr-test
+  namespace: openshift-machine-api
+spec:
+  type: reboot
+  machineName: test-machine
+```
+
+#### MachineRemediationRequest status API
+
+**MachineRemediationRequest** status will show what the state the remediation operation has and the time when the remediation operation started.
+
+```yaml
+status:
+  remediationState: InProgress
```

Comment:
Having "remediation" in these names makes them feel a bit redundant. What do you think about State and StartTime?

Reply:
We also need an `endTime`, probably a `reason` field (for additional information on failure), as well as a way to differentiate between "node is unhealthy but in a safe state to continue recovery" failed and a "we've lost any and all communication to / control over the node" failure.

Reply:
done